File: backend/app/services/order_display.py
"""Plain-language order and execution log labels for APIs and UI."""

from __future__ import annotations

from typing import Any

ORDER_STATUS_LABELS: dict[str, str] = {
    "paper_order_filled": "Filled at broker",
    "paper_order_rejected": "Paper order rejected",
    "paper_order_cancelled": "Cancelled",
    "paper_order_unfilled": "Not filled",
    "preflight_blocked": "Blocked by safety check before broker",
    "broker_rejected": "Broker rejected",
    "pending": "Pending",
    "submitted": "Sent to broker",
}

ORDER_TYPE_LABELS: dict[str, str] = {
    "marketable_limit_ioc": "Instant market-price limit order",
    "limit": "Limit order",
    "market": "Market order",
}

REJECTED_STATUSES = frozenset(
    {"paper_order_rejected", "preflight_blocked", "broker_rejected", "paper_order_cancelled", "paper_order_unfilled"}
)
FILLED_STATUSES = frozenset({"paper_order_filled", "filled"})
PREFLIGHT_STATUSES = frozenset({"preflight_blocked"})


def format_decimal(value: float | int | None, *, max_places: int = 8) -> str | None:
    if value is None:
        return None
    try:
        n = float(value)
    except (TypeError, ValueError):
        return str(value)
    if n == 0:
        return "0"
    text = f"{n:.{max_places}f}".rstrip("0").rstrip(".")
    if "." in text:
        whole, frac = text.split(".", 1)
        if len(frac) > 6:
            frac = frac[:6].rstrip("0")
        text = f"{whole}.{frac}" if frac else whole
    return text


def order_status_label(status: str | None) -> str:
    if not status:
        return "Unknown"
    return ORDER_STATUS_LABELS.get(status, status.replace("_", " ").capitalize())


def order_type_label(order_type: str | None) -> str:
    if not order_type:
        return "—"
    return ORDER_TYPE_LABELS.get(order_type, order_type.replace("_", " ").capitalize())


def reject_reason_plain(reason: str | None, *, status: str | None = None) -> str | None:
    if not reason:
        if status in PREFLIGHT_STATUSES:
            return "Blocked before broker — order was not sent to Alpaca."
        if status == "paper_order_rejected":
            return "Broker rejected this order after it was submitted."
        return None
    r = str(reason).strip()
    low = r.lower()
    if r == "STALE_QUOTE" or "stale_quote" in low:
        return (
            "Price quote was too old at submit time. Bot refreshed the quote; "
            "if still stale, it skipped instead of sending a bad order."
        )
    if "min_notional" in low or "notional" in low:
        return "Order size was below the broker minimum — rejected, not filled."
    if "insufficient" in low:
        return "Insufficient buying power or quantity — rejected, not filled."
    if "available" in low and "qty" in low:
        return "Broker reported zero available quantity to sell — rejected, not filled."
    if len(r) > 120:
        return r[:117] + "…"
    return r


def order_outcome_bucket(status: str | None, *, has_broker_id: bool = False) -> str:
    if not status:
        return "attempted"
    if status in PREFLIGHT_STATUSES:
        return "preflight_blocked"
    if status in REJECTED_STATUSES:
        return "rejected"
    if status in FILLED_STATUSES:
        return "filled"
    if has_broker_id or status in ("submitted", "pending"):
        return "sent"
    return "attempted"


def is_rejected_display(status: str | None) -> bool:
    return (status or "") in REJECTED_STATUSES


def _broker_detail_from_gates(gates_failed: dict) -> dict[str, Any]:
    if not isinstance(gates_failed, dict):
        return {}
    body = gates_failed.get("broker_error_body") or gates_failed.get("response_body")
    if body is None and gates_failed.get("broker"):
        import json

        raw = gates_failed.get("broker")
        if isinstance(raw, str) and raw.strip().startswith("{"):
            try:
                body = json.loads(raw)
            except json.JSONDecodeError:
                body = {"raw": raw}
        elif isinstance(raw, dict):
            body = raw
    return {
        "http_status": gates_failed.get("http_status"),
        "alpaca_code": gates_failed.get("alpaca_code"),
        "alpaca_message": gates_failed.get("alpaca_message") or gates_failed.get("broker_message"),
        "broker_error_body": body,
        "request_payload": gates_failed.get("request_payload"),
        "submitted_to_broker": gates_failed.get("submitted_to_broker"),
        "broker_response_received": gates_failed.get("broker_response_received"),
        "blocked_before_broker": gates_failed.get("blocked_before_broker"),
    }


def enrich_execution_row(row: dict[str, Any]) -> dict[str, Any]:
    status = row.get("status")
    broker_id = row.get("broker_order_id") or row.get("brokerOrderId")
    side = row.get("side") or ""
    gates_failed = row.get("gates_failed_json") or row.get("gates_failed") or {}
    broker_detail = _broker_detail_from_gates(gates_failed) if isinstance(gates_failed, dict) else {}
    if isinstance(gates_failed, dict) and gates_failed.get("outcome_bucket"):
        bucket = gates_failed.get("outcome_bucket")
    else:
        bucket = order_outcome_bucket(status, has_broker_id=bool(broker_id))
    rejected = is_rejected_display(status)
    broker_reject_after_submit = status == "paper_order_rejected" and (
        gates_failed.get("preflight_stage") == "broker_rejection"
        if isinstance(gates_failed, dict)
        else False
    )
    preflight = status in PREFLIGHT_STATUSES or (
        not broker_id
        and not broker_reject_after_submit
        and status in ("preflight_blocked", "paper_order_pending")
    )
    submitted = bool(broker_id) or broker_reject_after_submit or (
        isinstance(gates_failed, dict) and gates_failed.get("submitted_to_broker") is True
    )
    display_status = (
        "Blocked before broker"
        if preflight and not submitted
        else (
            "Broker rejected after submit"
            if broker_reject_after_submit
            else order_status_label(status)
        )
    )
    plain_reason = reject_reason_plain(row.get("reject_reason") or row.get("rejectReason"), status=status)
    if broker_reject_after_submit and broker_detail.get("alpaca_message"):
        plain_reason = str(broker_detail["alpaca_message"])
    return {
        **row,
        "status_label": display_status,
        "blocked_before_broker": preflight and not submitted,
        "submitted_to_broker": submitted,
        "broker_response_received": broker_detail.get("broker_response_received")
        if broker_detail
        else bool(broker_reject_after_submit),
        "broker_rejection": broker_detail if broker_detail else None,
        "alpaca_message": broker_detail.get("alpaca_message"),
        "order_type_label": order_type_label(row.get("order_type") or row.get("orderType")),
        "limit_price_display": format_decimal(row.get("limit_price") or row.get("limitPrice")),
        "filled_avg_price_display": format_decimal(row.get("filled_avg_price") or row.get("filledAvgPrice")),
        "requested_qty_display": format_decimal(row.get("requested_qty") or row.get("qty")),
        "reject_reason_plain": plain_reason,
        "outcome_bucket": bucket,
        "is_rejected": rejected,
        "is_filled": (status or "") in FILLED_STATUSES,
        "looks_like_closed_position": rejected and str(side).lower() == "sell",
        "user_message": (
            f"{str(side).upper()} {row.get('symbol', '?')}: {display_status}"
            + (f" — {plain_reason}" if rejected and plain_reason else "")
        ),
    }


def enrich_order_record(row: dict[str, Any]) -> dict[str, Any]:
    status = row.get("status")
    broker_id = row.get("broker_order_id") or row.get("brokerOrderId") or row.get("alpaca_order_id")
    bucket = order_outcome_bucket(status, has_broker_id=bool(broker_id))
    rejected = is_rejected_display(status) or "reject" in str(status or "").lower()
    filled = str(status or "").lower() == "filled"
    return {
        **row,
        "status_label": order_status_label(status) if status in ORDER_STATUS_LABELS else (
            "Filled" if filled else ("Rejected" if rejected else str(status or "Unknown").replace("_", " ").capitalize())
        ),
        "order_type_label": order_type_label(row.get("order_type") or row.get("orderType")),
        "qty_display": format_decimal(row.get("qty")),
        "filled_avg_price_display": format_decimal(row.get("filled_avg_price") or row.get("filledAvgPrice")),
        "outcome_bucket": bucket,
        "is_rejected": rejected,
        "is_filled": filled,
        "looks_like_closed_position": rejected and str(row.get("side", "")).lower() == "sell",
    }

File: src/lib/orderDisplay.ts
/** Plain-language order / execution log display (mirrors backend order_display.py). */

export type OrderOutcomeBucket =
  | "attempted"
  | "sent"
  | "filled"
  | "rejected"
  | "preflight_blocked";

const ORDER_STATUS_LABELS: Record<string, string> = {
  paper_order_filled: "Filled at broker",
  paper_order_rejected: "Paper order rejected",
  paper_order_cancelled: "Cancelled",
  paper_order_unfilled: "Not filled",
  preflight_blocked: "Blocked by safety check before broker",
  broker_rejected: "Broker rejected",
  pending: "Pending",
  submitted: "Sent to broker",
  filled: "Filled",
};

const ORDER_TYPE_LABELS: Record<string, string> = {
  limit_ioc: "IOC limit",
  marketable_limit_ioc: "IOC marketable limit",
  limit: "Limit",
  market: "Market",
};

/** First value that is not null/undefined/empty-string; else null. */
export function firstPresent(...vals: unknown[]): unknown {
  for (const v of vals) {
    if (v !== null && v !== undefined && v !== "") return v;
  }
  return null;
}

const REJECTED = new Set([
  "paper_order_rejected",
  "preflight_blocked",
  "broker_rejected",
  "paper_order_cancelled",
  "paper_order_unfilled",
]);

export function formatDecimal(value: unknown, maxPlaces = 8): string {
  if (value == null || value === "") return "—";
  const n = Number(value);
  if (Number.isNaN(n)) return String(value);
  if (n === 0) return "0";
  let text = n.toFixed(maxPlaces).replace(/\.?0+$/, "");
  const dot = text.indexOf(".");
  if (dot >= 0) {
    const whole = text.slice(0, dot);
    let frac = text.slice(dot + 1);
    if (frac.length > 6) frac = frac.slice(0, 6).replace(/0+$/, "");
    text = frac ? `${whole}.${frac}` : whole;
  }
  return text;
}

export function orderStatusLabel(status: string | null | undefined): string {
  if (!status) return "Unknown";
  return ORDER_STATUS_LABELS[status] ?? status.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}

export function orderTypeLabel(orderType: string | null | undefined): string {
  if (!orderType) return "—";
  return ORDER_TYPE_LABELS[orderType] ?? orderType.replace(/_/g, " ");
}

export function rejectReasonPlain(
  reason: string | null | undefined,
  status?: string | null
): string | null {
  if (!reason) {
    if (status === "preflight_blocked") return "Blocked by safety check before broker.";
    if (status === "paper_order_rejected") return "Broker or paper cage rejected this order — not filled.";
    return null;
  }
  const r = String(reason).trim();
  const low = r.toLowerCase();
  if (low.includes("min_notional") || low.includes("notional")) {
    return "Order size was below the broker minimum — rejected, not filled.";
  }
  if (low.includes("insufficient")) {
    return "Insufficient buying power or quantity — rejected, not filled.";
  }
  if (low.includes("available") && low.includes("qty")) {
    return "Broker reported zero available quantity to sell — rejected, not filled.";
  }
  return r.length > 120 ? `${r.slice(0, 117)}…` : r;
}

export function orderOutcomeBucket(
  status: string | null | undefined,
  hasBrokerId?: boolean
): OrderOutcomeBucket {
  if (!status) return "attempted";
  if (status === "preflight_blocked") return "preflight_blocked";
  if (REJECTED.has(status)) return "rejected";
  if (status === "paper_order_filled" || status === "filled") return "filled";
  if (hasBrokerId || status === "submitted" || status === "pending") return "sent";
  return "attempted";
}

export function isRejectedStatus(status: string | null | undefined): boolean {
  return !!status && (REJECTED.has(status) || status.toLowerCase().includes("reject"));
}

export function enrichExecutionRow(row: Record<string, unknown>): Record<string, unknown> {
  const status = String(row.status ?? "");
  const brokerId = row.broker_order_id ?? row.brokerOrderId;
  const side = String(row.side ?? "");
  const rejected = isRejectedStatus(status);
  const bucket = orderOutcomeBucket(status, !!brokerId);
  const reason = (row.reject_reason ?? row.rejectReason) as string | undefined;
  return {
    ...row,
    status_label: orderStatusLabel(status),
    order_type_label: orderTypeLabel(String(row.order_type ?? row.orderType ?? "marketable_limit_ioc")),
    limit_price_display: formatDecimal(row.limit_price ?? row.limitPrice),
    filled_avg_price_display: formatDecimal(row.filled_avg_price ?? row.filledAvgPrice),
    requested_qty_display: formatDecimal(row.requested_qty ?? row.qty),
    reject_reason_plain: rejectReasonPlain(reason, status),
    outcome_bucket: bucket,
    is_rejected: rejected,
    looks_like_closed_position: rejected && side.toLowerCase() === "sell",
  };
}

export function enrichOrderRecord(row: Record<string, unknown>): Record<string, unknown> {
  const status = String(firstPresent(row.status, row.broker_status, row.outcome) ?? "");
  const brokerId = row.broker_order_id ?? row.brokerOrderId ?? row.alpaca_order_id;
  const rejected = isRejectedStatus(status) || status.toLowerCase().includes("reject");
  const lower = status.toLowerCase();
  const filled = lower === "filled" || lower === "paper_order_filled";
  const side = String(row.side ?? "");
  // API rows expose type/filled_qty/requested_qty (not order_type/qty) — map via fallback chains.
  const orderType = firstPresent(row.order_type, row.type, row.orderType, row.order_class);
  const qtyVal = firstPresent(row.filled_qty, row.qty, row.requested_qty, row.submitted_qty, row.broker_qty);
  const priceVal = firstPresent(
    row.filled_avg_price,
    row.avg_fill_price,
    row.avg_entry,
    row.limit_price,
    row.current_price
  );
  const timeVal = firstPresent(row.filled_at, row.submitted_at, row.created_at, row.timestamp);
  return {
    ...row,
    status_label: filled ? "Filled" : rejected ? "Rejected" : orderStatusLabel(status),
    order_type_label: orderType ? orderTypeLabel(String(orderType)) : "—",
    qty_display: formatDecimal(qtyVal),
    requested_qty_display: formatDecimal(qtyVal),
    filled_avg_price_display: formatDecimal(priceVal),
    limit_price_display: formatDecimal(firstPresent(row.limit_price, row.limitPrice)),
    time_display: timeVal != null ? String(timeVal) : "—",
    outcome_bucket: orderOutcomeBucket(status, !!brokerId),
    is_rejected: rejected,
    looks_like_closed_position: rejected && side.toLowerCase() === "sell",
  };
}

export interface OrderSummaryCounts {
  orders_attempted?: number;
  orders_sent_to_broker?: number;
  orders_filled?: number;
  orders_rejected?: number;
  orders_blocked_preflight?: number;
  last_order_user_message?: string;
}
